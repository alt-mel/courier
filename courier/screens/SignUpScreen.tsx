import React, { useState } from 'react';
import {
  Pressable,
  ActivityIndicator,
  Alert,
} from 'react-native';
import { View, Text, TextInput } from '../components/Themed';
import { useLinkTo } from '@react-navigation/native';

import AsyncStorage from '@react-native-async-storage/async-storage';

import { useMutation, gql } from '@apollo/client';

const SIGN_UP_MUTATION = gql`
  mutation signUp($email: String!, $password: String!, $name: String!) {
    signUp(input: { email: $email, password: $password, name: $name }) {
      token
      user {
        id
        name
        email
      }
    }
  }
`;

const SignUpScreen = ({ navigation }) => {
  const [email, setEmail] = useState('');
  const [name, setName] = useState('');
  const [password, setPassword] = useState('');
  const linkTo = useLinkTo();

  const [signUp, { data, error, loading }] = useMutation(SIGN_UP_MUTATION);

  if (error) {
    Alert.alert('Error signing up. Try again');
  }

  if (data) {
    AsyncStorage.setItem('token', data.signUp.token).then(() => {
      // navigation.navigate('Home');
      linkTo('/Home/Deliveries');
    });
  }

  const onSubmit = () => {
    signUp({ variables: { name, email, password } });
  };

  return (
    <View style={{ padding: 20 }}>
      <TextInput
        placeholder="name"
        value={name}
        onChangeText={setName}
        style={{
          fontSize: 18,
          width: '100%',
          marginVertical: 25,
        }}
      />

      <TextInput
        placeholder="email@email.com"
        value={email}
        onChangeText={setEmail}
        style={{
          fontSize: 18,
          width: '100%',
          marginVertical: 25,
        }}
      />

      <TextInput
        placeholder="password"
        value={password}
        onChangeText={setPassword}
        secureTextEntry
        style={{
          fontSize: 18,
          width: '100%',
          marginVertical: 25,
        }}
      />

      <Pressable
        onPress={onSubmit}
        style={{
          backgroundColor: '#ed706E',
          height: 50,
          borderRadius: 5,
          alignItems: 'center',
          flexDirection: 'row',
          justifyContent: 'center',
          marginTop: 30,
        }}
      >
        {loading && <ActivityIndicator />}
        <Text
          style={{
            color: 'white',
            fontSize: 18,
            fontWeight: 'bold',
          }}
        >
          Sign up
        </Text>
      </Pressable>

      <Pressable
        onPress={() => {
          navigation.navigate('SignIn');
        }}
        disabled={loading}
        style={{
          height: 50,
          borderRadius: 5,
          alignItems: 'center',
          justifyContent: 'center',
          marginTop: 30,
        }}
      >
        <Text
          style={{
            color: '#ed706E',
            fontSize: 18,
            fontWeight: 'bold',
          }}
        >
          Already Have an account? Sign in
        </Text>
      </Pressable>
    </View>
  );
};

export default SignUpScreen;
