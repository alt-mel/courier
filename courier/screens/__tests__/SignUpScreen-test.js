import * as React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import {render, fireEvent, act } from '@testing-library/react-native';
import wait from "waait";

import SignUpScreen, { SIGN_UP_MUTATION } from '../SignUpScreen';

const mocks = [ {
    request: {
      query: SIGN_UP_MUTATION,
      variables: {
        name: 'name',
        email: 'email',
        password: 'password'
      },
    },
    result: {
      data: {
        signUp: {
          token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjYxODU0NzE1YjExYjExNzJkZjY2M2RhNyIsImlhdCI6MTYzNjEyNDQzNywiZXhwIjoxNjM4NzE2NDM3fQ.l_DoOqDVYwSycLl4dovEn1ypIM71kYWTzU-uLKT9f3s',
          user: {
            id: '61363974b6f26c6cd7881183',
            name: 'Username',
            email: 'admin@john.com'
          }
        }
      }
    },
  },];

  it('renders the Sign Up Screen', () => {
    const tree = render( <MockedProvider><SignUpScreen/></MockedProvider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it('renders default elements', () => {
    const { getByPlaceholderText, getByTestId } = render(<MockedProvider mocks={mocks} addTypename={false}><SignUpScreen/></MockedProvider> );
    expect(getByPlaceholderText("name")).not.toBeNull();
    expect(getByPlaceholderText("email@email.com")).not.toBeNull();
    expect(getByPlaceholderText("password")).not.toBeNull();
    expect(getByTestId('SignUp.Button'))
  });

  it('handles valid input submission', async () => {
    const pushMock = jest.fn();
    const {getByTestId } = render(<MockedProvider mocks={mocks} addTypename={false}><SignUpScreen navigation={{ navigate: pushMock }}/></MockedProvider> );
  
    fireEvent.changeText(getByTestId('SignUp.Name'), 'name'),
    fireEvent.changeText(getByTestId('SignUp.Email'), 'email'),
    fireEvent.changeText(getByTestId('SignUp.Password'), 'password'),    
    fireEvent.press(getByTestId('SignUp.Button'))
  
    await act(async () => {
      await wait(0);
    });
  
    expect(pushMock).toBeCalledWith("Home")
  });