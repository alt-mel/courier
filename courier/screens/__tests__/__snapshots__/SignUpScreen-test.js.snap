// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Sign Up Screen 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "padding": 20,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="name"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "fontSize": 18,
          "marginVertical": 25,
          "width": "100%",
        },
      ]
    }
    testID="SignUp.Name"
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="email@email.com"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "fontSize": 18,
          "marginVertical": 25,
          "width": "100%",
        },
      ]
    }
    testID="SignUp.Email"
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "fontSize": 18,
          "marginVertical": 25,
          "width": "100%",
        },
      ]
    }
    testID="SignUp.Password"
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#bebebe",
        "borderRadius": 5,
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "center",
        "marginTop": 30,
      }
    }
    testID="SignUp.Button"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "color": "white",
            "fontSize": 18,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Sign up
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 5,
        "height": 50,
        "justifyContent": "center",
        "marginTop": 30,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "color": "#bebebe",
            "fontSize": 18,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Already Have an account? Sign in
    </Text>
  </View>
</View>
`;
